Added a layer of self description for each Era

TimSheard opened the pull request:
When one imports one of the Era modules qualified one has access to a number of properties of that Era. For example

import qualified Cardano.Ledger.Shelley as Shelley

then 
Shelley.Tx   is the Tx for the Shelley era, this holds for all the Core.* type families
Shelley.Era is the Era for Shelley i.e. ShelleyEra c
Shelly.proxy is a proxy for that Era. etc.

In the file Cardano.Ledger.DescribesEra are a set of assertions of what properties hold for each era, and some tests that they actually do. One can see how an Era is constructed by looking at its description. For example here is the description for Mary.

type DescribesMary era =
  ( WellFormed era,
    Core.Value era ~ Mary.Value era,
    Core.TxBody era ~ Allegra.TxBody era,
    Core.TxOut era ~ Shelley.TxOut era,
    Core.Script era ~ Allegra.Script era,
    Core.AuxiliaryData era ~ Allegra.AuxiliaryData era,
    Core.PParams era ~ Shelley.PParams era,
    Core.Tx era ~ Shelley.Tx era
  )

TimSheard committed:
Added self-describing features to Era instances.

TimSheard committed:
Added tests, that test each era against its description.

TimSheard committed:
ormolise

JaredCorduan reviewed (approved):
I love it!

nc6 commented on alonzo/impl/cardano-ledger-alonzo.cabal:
```diff
@@ -20,9 +20,10 @@ library
   if impl(ghc < 8.10)
     buildable: False
   exposed-modules:
+    Cardano.Ledger.DescribeEras
```
This isn't to do with Alonzo. However, there is a new test package `cardano-ledger-test`, which sits atop all eras and is intended to be the home for new tests that use all eras - this would be a perfect fit for that package.

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo.hs:
```diff
@@ -58,3 +71,10 @@ instance
+proxy :: Proxy (AlonzoEra c)
+proxy = Proxy
+
+type Era c = AlonzoEra c
```
This is a very confusing export. We use `Era` all over the place, it always refers to the Era class. I think it's a bad idea to introduce this alias.

nc6 commented on alonzo/impl/src/Cardano/Ledger/DescribeEras.hs:
```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Ledger.DescribeEras
```
As mentioned, move this to `cardano-ledger-test` package

nc6 commented on alonzo/impl/test/test/Test/Cardano/Ledger/Alonzo/DescribeEra.hs:
```diff
@@ -0,0 +1,22 @@
+module Test.Cardano.Ledger.Alonzo.DescribeEra (describeTest) where
```
Likewise move to `cardano-ledger-test`, remove reference to Alonzo in module name.

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/Allegra.hs:
```diff
@@ -61,3 +84,14 @@ instance
+proxy :: Proxy (ShelleyMAEra 'Allegra c)
+proxy = Proxy
+
+type Era c = ShelleyMAEra 'Allegra c
```
Ditto on this alias

nc6 commented on shelley-ma/impl/src/Cardano/Ledger/Mary.hs:
```diff
@@ -63,3 +85,16 @@ instance
+proxy :: Proxy (ShelleyMAEra 'Mary c)
+proxy = Proxy
+
+type Era c = ShelleyMAEra 'Mary c
```
And tritto. Is that a word?

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Shelley.hs:
```diff
@@ -86,3 +97,24 @@ instance CryptoClass.Crypto c => ValidateAuxiliaryData (ShelleyEra c) where
+
+type Value era = Coin
+
+proxy :: Proxy (ShelleyEra c)
```
I would argue another somewhat confusing export.

nc6 reviewed (changes requested):
I like the idea of qualified imports gathering all the various types! But have requested a few updates:

- The `DescribeEra` stuff would be a much better fit for the `cardano-ledger-test` package.
- The `Era` synonym is just going to be confusing.
- The `proxy` synonym arguably also so?

erikd commented:
The `Evidence` and `Witness` bits of this look like they would make my life immensely easier.

erikd commented:
I have cargo culted a bit of this into `db-sync` i would like to suggest that the name of the `Witness` constructors change to `WitnessShelley`, `WitnessAllegra` and so on to avoid name clashes.

TimSheard committed:
Addressed comment by Nick.

Moved the actual test file DescribeEra.hs to cardano-ledger-test/src/Test/Cardano/Ledger/DescribeEra.hs
  Not sure how to run this test from there.
Moved Cardano/Crypto/VRF/Fake.hs to shelley/chain-and-ledger/executable-spec/src/Cardano/Crypto/VRF/Fake.hs
  This suports defining both StandardCrypto and TestCrypto in  Cardano.Ledger.Crypto
For each new Era XXX, XXX.Era has been rename to XXX.Self, e.g. Shelley.SheleyEra and Shelley.Self are the same type.
Removed the proxy type synonym.

nc6 commented:
Sounds like the current plan is not to merge this
